### DISCUSSION TITLE
Display skeleton rows during initial load

### DISCUSSION THREAD

**Discussion by taschetto** [TanStack/table#2386] [General] [answered]
Hi! I want to display an arbitrary number of rows whilst fetching the first page. Here's what I got:

![skeleton](https://user-images.githubusercontent.com/5279182/83207671-26ceb200-a12a-11ea-8fb3-5906cbd9612f.gif)

Visually it is what I want. But the code doesn't seem any good.

```javascript
{loading &&
  Array.from({ length: skeletonRows }).map(() => {
    return headerGroups.map(headerGroup => (
      <tr {...headerGroup.getHeaderGroupProps()}>
        {headerGroup.headers.map(column => (
          <td {...column.getHeaderProps()}>
            <Skeleton />
          </td>
        ))}
      </tr>
    ))
  })}
```

Does anyone have a better approach to this?

**Comment by GuptaSiddhant** [comment, marked as answer]
Hi,

Recently I had added Skeleton loading to my table. First I attempted what you're doing but I quickly went ahead with other approach.

The idea is to change the data & columns array passed to useTable if the content is loading. The following code create 30 rows (you can change according to your requirement) with empty data, and replace each cell with Skeleton component. 
The good thing is your column headers/footers are intact.

```js
const Table = (columns, data) => {
  ...
  const tableData = React.useMemo(
    () => (loading ? Array(30).fill({}) : data),
    [loading, data]
  ); 
  const tableColumns = React.useMemo(
    () =>
      loading
        ? columns.map((column) => ({
            ...column,
            Cell: <Skeleton />,
          }))
        : columns,
    [loading, columns]
  );

  useTable({ columns: tableColumns, data: tableData });
 ...
}
```

**Reply by taschetto** [reply]
Very clever and seemly the right way to do it. Thank you!

**Reply by tyteen4a03** [reply]
This solution seems to kill performance for some reason. For context, I am using this with pagination and every time I change pages the page just stops responding.

**Reply by GuptaSiddhant** [reply]
@tyteen4a03 I tested the solution with client-side pagination and I couldn't notice any lag or delay. If you're using server-side pagination, then you can share a codesandbox link and maybe we can figure it out.
Also, I've kept 100 rows per page as default.

**Reply by taschetto** [reply]
I'm using it with server-side pagination without issues.

Also, from a UX perspective, I think the Skeleton UI should appear only before and during the first fetch, when you have no data to display. Then, when you are currently showing some rows containing real data, you shouldn't use the Skeleton - only a loading indicador.

**Reply by laurent512** [reply]
> Hi,
> 
> Recently I had added Skeleton loading to my table. First I attempted what you're doing but I quickly went ahead with other approach.
> 
> The idea is to change the data & columns array passed to useTable if the content is loading. The following code create 30 rows (you can change according to your requirement) with empty data, and replace each cell with Skeleton component.
> The good thing is your column headers/footers are intact.
> 
> ```js
> const Table = (columns, data) => {
>   ...
>   const tableData = React.useMemo(
>     () => (loading ? Array(30).fill({}) : data),
>     [loading, data]
>   ); 
>   const tableColumns = React.useMemo(
>     () =>
>       loading
>         ? columns.map((column) => ({
>             ...column,
>             Cell: <Skeleton />,
>           }))
>         : columns,
>     [loading, columns]
>   );
> 
>   useTable({ columns: tableColumns, data: tableData });
>  ...
> }
> ```

Hello, thank you for the nice answer ! 
Do you know how to to if I have columns inside columns ? e.g.

```
FullName
--------------------------
First Name | Last Name
```

Because the map functions inside tableColumns  will only "FullName" and not its sub columns.

thank you so much

**Reply by GuptaSiddhant** [reply]
@laurent512 It is interesting scenario once I didn’t test with grouped headers. I was under the assumption that since I’m overriding the “Cell” property, the “Header” property should remain intact.

**Reply by laurent512** [reply]
I am trying to map inside map but couldn't make it work ( I am quite new in js development)

I achieved to do it with a quite ugly workaround  (knowing I have 3 "Top" columns) in which I go through with your code: 

```
   const isLoadingColumns = [ 
        {
            Header: columns[0].Header,
            columns: columns[0].columns.map((column) => ({
                ...column,
                Cell: 'loading'
            }))
        },
        {
            Header: columns[1].Header,
            columns: columns[1].columns.map((column) => ({
                ...column,
                Cell: 'loading'
            }))
        },
        {
            Header: columns[2].Header,
            columns: columns[2].columns.map((column) => ({
                ...column,
                Cell: 'loading'
            }))
        }
    ];
```

If you have something cleaner in mind, I will be eternally grateful  ^^

**Reply by ghost** [reply]
I keep getting a forEach error with the provided solution above. Not sure why?

**Reply by isocroft** [reply]
This answer didn't work for me. Instead of returning the `<Skeleton />`,in the table cell, it returned `undefined`.

**Reply by Skia69** [reply]
> This answer didn't work for me. Instead of returning the `<Skeleton />`,in the table cell, it returned `undefined`.

Same goes here

The cells returned are undefined and i have no clue how to fix it

**Reply by tacosjs** [reply]
use `cell` with a function return instead of `Cell`

```js
const columnsMemo = useMemo(
    () =>
      isLoading
        ? columns.map((column) => ({
            ...column,
            cell: () => <Skeleton />,
          }))
        : columns,
    [isLoading, columns]
  );
```

**Reply by jsaunders92** [reply]
I don't suppose anyone has an example of this by any chance?

**Reply by mbyrne00** [reply]
Works a treat for me ... for those looking for the `Skeleton` component it's in [mui](https://mui.com/material-ui/react-skeleton/). But you can swap out for your favourite component. The concept is the same ... fill x number of dummy rows and make sure every col definition has a `cell` function defined that returns your placeholder element.

Thanks @GuptaSiddhant

**Reply by JulienLemaitre** [reply]
If you're using [Chakra UI](https://v2.chakra-ui.com/docs/components/skeleton), add a `height` to the `Skeleton` component.
```javascript
cell: () => <Skeleton height="20px" />,
```

**Comment by sagar-jani** [comment]
Unfortunately this give me issues with `{table.getHeaderGroups().map((headerGroup) => (` code. 

Here is how I'm using with ShadCN

```
export function DataTableDemo({ data, loading }) {
  console.log('data', data);
  const [sorting, setSorting] = React.useState([]);
  const [columnFilters, setColumnFilters] = React.useState([]);
  const [columnVisibility, setColumnVisibility] = React.useState({});
  const [rowSelection, setRowSelection] = React.useState({});

  const tableData = React.useMemo(
    () => (loading ? Array(30).fill({}) : data),
    [loading, data]
  );
  const tableColumns = React.useMemo(
    () =>
      loading
        ? columns.map((column) => ({
            ...column,
            Cell: () => (
              <Skeleton className="h-[14px] w-[60%] rounded-sm absolute left-3 top-[39px]" />
            )
          }))
        : columns,
    [loading]
  );

  const table = useReactTable({
    tableData,
    tableColumns,
    onSortingChange: setSorting,
    onColumnFiltersChange: setColumnFilters,
    getCoreRowModel: getCoreRowModel(),
    getPaginationRowModel: getPaginationRowModel(),
    getSortedRowModel: getSortedRowModel(),
    getFilteredRowModel: getFilteredRowModel(),
    onColumnVisibilityChange: setColumnVisibility,
    onRowSelectionChange: setRowSelection,
    state: {
      sorting,
      columnFilters,
      columnVisibility,
      rowSelection,
      getPaginationRowModel: getPaginationRowModel()
    }
  });
```

**Reply by dmitry-mitrakhovich** [reply]
```
 const tableColumns = React.useMemo(
    () =>
      loading
        ? columns.map((column) => ({
            ...column,
            cell: () => (
              <Skeleton className="h-[14px] w-[60%] rounded-sm absolute left-3 top-[39px]" />
            )
          }))
        : columns,
    [loading]
  );
```

Use `cell` instead `Cell`

**Reply by dmnlgn** [reply]
and you should change `tableData` to `data: tableData`, and `tableColumns` to `columns: tableColumns`

**Comment by Dezmonter** [comment]
```
const tableColumns = useMemo<ColumnDef<TData, TValue>[]>(
() =>
	loading
		? columns.map((column) => {
			if ('columns' in column) {
				return {
					...column,
					columns: column.columns?.map((nestedColumn) => ({
						...nestedColumn,
						cell: () => (
							<Skeleton className="h-[20px] w-[80%] rounded-lg" />
						),
					})),
				}
			}

			return {
				...column,
				cell: () => (
					<Skeleton className="h-[20px] w-[80%] rounded-lg" />
				),
			}
		})
		: columns,
[loading]
)
```


This is my solution for working with grouped columns

**Comment by fblettner** [comment]
I did this in the table body. Keep Table header as it is and add condition for loading only into the body : 

```typescript
isLoading
                            ? (
                                <TableBody>
                                    {renderSkeletonRows()}
                                </TableBody>)
                            : table.getRowModel().rows.length === 0
                                ? null
                                : (
                                    <LYTableBody
                                        rowVirtualizer={rowVirtualizer}
                                        table={table}
                                        tableState={tableState}
                                        tableRef={tableRef}
                                        onSelect={onSelect}
                                        onMouseDown={onMouseDown}
                                        onTouchStart={onTouchStart}
                                        onTouchEnd={onTouchEnd}
                                        onDoubleClick={onDoubleClick}
                                    />
                                )
```

and renderSkeletonRows is this one : 
```typescript
    const renderSkeletonRows = () => (
        <Fragment>
            {/* Render 10 skeleton rows as a placeholder while loading */}
            {Array.from(new Array(10)).map((_, index) => (
                <TableRow key={index}>
                    {/* Render skeleton cells for each visible column */}
                    {table.getVisibleLeafColumns().map((column) => (
                        <TableCell key={column.id}>
                            <Skeleton variant="rounded" width="100%" height={8} />
                        </TableCell>
                    ))}
                </TableRow>
            ))}
        </Fragment>
    );
```

Table and Skeleton are custom components in my framework because I have completely remove any MUI dependencies

You can find full code for this here : https://github.com/fblettner/liberty/blob/main/frontend/src/components/forms/FormsTable/views/TableGrid.tsx
